pennylane/workflow/construct_batch.py

```diff
@@ -81,6 +83,105 @@ def _get_full_transform_program(
     return program + qnode.device.preprocess_transforms(config)
 
 
+def _interpret_level_initial(
+    level: Optional[Literal["top", "user", "device", "gradient"] | int | slice],
+    num_user_transforms: int,
+) -> slice:
+    """Interpret the level specification for the initial user transform slice.
+
+    This function handles slicing into the user transforms before any
+    gradient or device transforms are applied.
+
+    Args:
+        level: The level specification from user input
+        num_user_transforms: Number of user transforms
+
+    Returns:
+        slice: The slice to apply to the user transform program
+    """
+    if level == "top":
+        return slice(0, 0)
+
+    if level in ("user", "device", "gradient"):
+        return slice(0, num_user_transforms)
+
+    if isinstance(level, str):  # pragma: no cover
+        raise ValueError(
+            f"level {level} not recognized. Acceptable strings are 'device', 'top', 'user', and 'gradient'."
+        )
+
+    if level is None:
+        return slice(0, num_user_transforms)
+
+    if isinstance(level, int):
+        return slice(0, min(level, num_user_transforms))
+
+    # Handle slice objects - clamp to user transform bounds
+    if isinstance(level, slice):
+        start = level.start if level.start is not None else 0
+        return slice(start, level.stop, level.step)
+
+    return level  # pragma: no cover
+
+
+# pylint: disable=too-many-return-statements
+def _interpret_level_inner(
+    level: Optional[Literal["top", "user", "device", "gradient"] | int | slice],
+    num_user_transforms: int,
+    has_gradient_expand: bool,
+    has_final_transform: bool = False,
+) -> slice:
+    """Interpret the level specification for the inner transform slice.
+
+    This function handles slicing into the remaining transforms (gradient + device)
+    after user transforms have already been applied. The returned slice is already
+    shifted to account for final transforms being moved to the end.
+
+    Args:
+        level: The level specification from user input
+        num_user_transforms: Number of user transforms (already applied)
+        has_gradient_expand: Whether gradient expansion transform exists
+        has_final_transform: Whether there's a final transform that gets moved to the end
+
+    Returns:
+        slice: The slice to apply to the remaining transform program (already shifted)
+    """
+    # Calculate the shift needed to account for final transforms
+    shift_len = num_user_transforms - int(has_final_transform)
+    
+    def _shift(i):
+        return i - shift_len if i is not None else i
+
+    start = max(0, num_user_transforms)
+    if level in ("top", "user"):
+        return slice(_shift(start), _shift(0))  # No additional transforms needed
```

Comment:
Would we even hit this? Maybe for safety:
```suggestion
    if level in ("top", "user"):
        raise ValueError("We should not be able to hit this logical branch")
```